Userland/Libraries/LibSQL/SQLClient.cpp

```diff
@@ -5,10 +5,132 @@
+
+// This is heavily based on how SystemServer's Service creates its socket.
+static ErrorOr<int> create_database_socket(DeprecatedString const& socket_path)
+{
+    if (Core::File::exists(socket_path))
+        TRY(Core::System::unlink(socket_path));
+
+#    ifdef SOCK_NONBLOCK
+    auto socket_fd = TRY(Core::System::socket(AF_LOCAL, SOCK_STREAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0));
+#    else
+    auto socket_fd = TRY(Core::System::socket(AF_LOCAL, SOCK_STREAM, 0));
+
+    int option = 1;
+    TRY(Core::System::ioctl(socket_fd, FIONBIO, &option));
+    TRY(Core::System::fcntl(socket_fd, F_SETFD, FD_CLOEXEC));
+#    endif
+
+    TRY(Core::System::fchmod(socket_fd, 0600));
+
+    auto socket_address = Core::SocketAddress::local(socket_path);
+    auto socket_address_un = socket_address.to_sockaddr_un().release_value();
+
+    TRY(Core::System::bind(socket_fd, reinterpret_cast<sockaddr*>(&socket_address_un), sizeof(socket_address_un)));
+    TRY(Core::System::listen(socket_fd, 16));
+
+    return socket_fd;
+}
+
+static ErrorOr<void> launch_server(DeprecatedString const& socket_path, DeprecatedString const& pid_path, char const* server_path)
+{
+    auto server_fd = TRY(create_database_socket(socket_path));
+    auto server_pid = TRY(Core::System::fork());
+
+    if (server_pid == 0) {
+        TRY(Core::System::setsid());
+        TRY(Core::System::signal(SIGCHLD, SIG_IGN));
+        server_pid = TRY(Core::System::fork());
+
+        if (server_pid != 0) {
+            auto server_pid_file = TRY(Core::Stream::File::open(pid_path, Core::Stream::OpenMode::Write));
+            TRY(server_pid_file->write(DeprecatedString::number(server_pid).bytes()));
+
+            exit(0);
+        }
+
+        server_fd = TRY(Core::System::dup(server_fd));
+
+        auto takeover_string = DeprecatedString::formatted("{}:{}", socket_path, server_fd);
+        TRY(Core::System::setenv("SOCKET_TAKEOVER"sv, takeover_string, true));
+
+        char const* argv[] = {
+            "SQLServer",
+            "--pid-file",
+            pid_path.characters(),
+            nullptr,
+        };
+
+        if (execvp(server_path, const_cast<char**>(argv)) < 0)
+            perror("execvp");
+
+        TRY(Core::System::unlink(pid_path));
+        VERIFY_NOT_REACHED();
+    }
+
+    TRY(Core::System::waitpid(server_pid));
+    return {};
+}
+
+static ErrorOr<bool> should_launch_server(DeprecatedString const& pid_path)
+{
+    if (!Core::File::exists(pid_path))
+        return true;
+
+    Optional<pid_t> pid;
+    {
+        auto server_pid_file = TRY(Core::Stream::File::open(pid_path, Core::Stream::OpenMode::Read));
+        auto contents = TRY(server_pid_file->read_all());
+        pid = StringView { contents }.to_int<pid_t>();
+    }
+
+    if (!pid.has_value()) {
+        dbgln("SQLServer PID file '{}' exists, but with an invalid PID", pid_path);
+        TRY(Core::System::unlink(pid_path));
+        return true;
+    }
+    if (kill(*pid, 0) < 0) {
```

Comment:
This could also fail if the pid_path is owned by another user or group. In that case we'd probably fail to unlink though?

Reply:
Yeah we get a permission denied error:
```
Runtime error: open: Permission denied (errno=13)
```

That's not very helpful though, nothing indicating what it actually failed to open. So I'll add a check here to at least get a nicer error message.